### ISSUE TITLE
Unidirectional Correspondences question

### ISSUE THREAD

**Issue by AliYoussef97** [facebookresearch/silk#50]
Hello!

I had a question regarding the unidirectional correspondences function [here](https://github.com/facebookresearch/silk/blob/aa51cd46899346cfa2873bbc06d6e11c6b9d92f7/lib/losses/info_nce/loss.py#L13C1-L47).

I was recreating a part of the code, and I am getting different results in that exact function depending on the ordering of the points.

Case1: I initially create a mesh_grid but I use [ys,xs] instead of [xs,ys], when I am applying the transformation, I flip the points to [xs,ys] and after the transformation I convert back to [ys,xs], when passing the points to unidirectional correspondence function, I choose the ordering as "yx" as I am using row-major order. However this results to corr_0 and corr_1 as vectors filled with -1.

Case2: If I create a mesh_grid as [xs,ys], apply the same transformation and pass the points to unidirectional coresspondence function with "xy" ordering, then everything works as intended.

I do not fully understand the reason for this, maybe I am missing something or I misunderstood the ordering. I checked Case1 and Case2 points simulatenously and they align, the only difference is Case1 is row-major order and Case2 is column-major order. 

I also had another question regadring coord_mapping, I made a small figure for simplicity to check if my understanding is correct?

![mapping](https://github.com/facebookresearch/silk/assets/122641881/57f970ad-4182-4694-b6d0-c8dbe9160973)

Thank you!

**Comment by gleize** [comment]
Hi @AliYoussef97,

I've added some test to double-check the ordering of the `positions_to_unidirectional_correspondence` function (c.f. [here](https://github.com/facebookresearch/silk/pull/51/commits/23eb33d705a61cdeca8be83d57cb04d7187c87fd)). It works as expected.

It's difficult to figure out exactly what's wrong without seeing the code, so I can only speculate.
* First, double-check that you are feeding the width and height in the same order to the `positions_to_unidirectional_correspondence` function. It should not be flipped between case 1 and 2. Only the x,y dimensions in `positions` should be flipped.
* Second, verify that the xy-positions and yx-positions you feed to `positions_to_unidirectional_correspondence` in both cases are exactly the same (up to flipping and maybe small floating error). If that's not the case, then I suspect there could be an issue with how you apply the transform. If done incorrectly, it could send all the points to be out-of-bound, which would result in `-1` everywhere.

> I also had another question regadring coord_mapping, I made a small figure for simplicity to check if my understanding is correct?

That's correct, except I think you made a mistake on the `Image Coordinates`. It should be `(222, 302)` instead of `(231, 311)`. The 9-pixels no-detection border is removed from all sides.

A nice way to visualize this mapping is to think of each feature position as being in the center of a succession of convolutions. Those aggregated convolutions represent a block in image space (i.e. receptive field) of size 2x9 + 1.

**Comment by AliYoussef97** [comment]
Hello @gleize 

Thank you so much!

> That's correct, except I think you made a mistake on the `Image Coordinates`. It should be `(222, 302)` instead of `(231, 311)`. The 9-pixels no-detection border is removed from all sides.
>
> A nice way to visualize this mapping is to think of each feature position as being in the center of a succession of convolutions. Those aggregated convolutions represent a block in image space (i.e. receptive field) of size 2x9 + 1.

Apologies if the figure was incorrect regarding the image coordinates, my intention was that the sparse interest points are shifted by +9 pixels to go from feature map coordinates to image coordinates, however from my understanding the shape of the map itself is still the same. This assumeption is based on the [LinearCoordinateMapping ](https://github.com/facebookresearch/silk/blob/aa51cd46899346cfa2873bbc06d6e11c6b9d92f7/lib/backbones/silk/coords.py#L29C1-L66C1) class, which default has a scale of 1. and a bias of 9.  However I am not completely sure where the +1 originates from in the 2x9 + 1. 

&nbsp;

> It's difficult to figure out exactly what's wrong without seeing the code, so I can only speculate.
> - First, double-check that you are feeding the width and height in the same order to the > positions_to_unidirectional_correspondence function. It should not be flipped between case 1 and 2. Only the x,y dimensions in positions should be flipped.
> -Second, verify that the xy-positions and yx-positions you feed to positions_to_unidirectional_correspondence in both cases are exactly the same (up to flipping and maybe small floating error). If that's not the case, then I suspect there could be an issue with how you apply the transform. If done incorrectly, it could send all the points to be out-of-bound, which would result in -1 everywhere.


Very sorry for not being thorough in my initial post. I follow positions_to_unidirectional_correspondence and keep_mutual_correspondences_only, however, I do not use homography, instead I use projective prespective geomerty.

For example, starting with an input Image of size (480,640), the feature map would be size (462,622), I create a row-major order grid map (y,x), I add the +9 bias to convert back to image coordinates, I then project the grid map points from Camera 1 Frame to Camera 2 frame (I use that as corr_forward) and from Camera 2 to 1 (corr_backward). The following figure would be an example of the corr_forward and corr_backward

I then follow positions_to_unidirectional_correspondence and keep_mutual_correspondences_only, my issue is:

![Example](https://github.com/facebookresearch/silk/assets/122641881/940aa8be-3958-4209-b0e3-95ac9316b0ec)


- When I sample the grid as (y,x), I flip to (x,y) during prespective projection then flip to (y,x) and during positions_to_unidirectional_correspondence I set "ordering" as "yx", this results in a vector of -1s.

- If I sample gris as (x,y), I directly do prespective projection without flipping, then set ordering as "xy" in positions_to_unidirectional_correspondence, then I get a corresponce of 220k points out of 287364. 

In both cases above, I have checked the resulting tensors after projection, and they are exactly the same only difference is row or column major order. However, seeing the test you've conducted, I realise it has to be a mistake from my end. Extremely sorry for taking your time in that manner.

Thank you again! :D

**Comment by gleize** [comment]
> I am not completely sure where the +1 originates from in the 2x9 + 1

If you take a single 3x3 conv layer with no padding, the "removed" border is of size 1 (the 9 in your case), which correspond to (K - 1) / 2. K being the kernel size.
So, once we know the padding size (9), we can infer the size of the receptive field (or "kernel") by inverting this formula, thus the formulation above.

> Extremely sorry for taking your time in that manner.

No worry. One of the ordering wasn't tested initially, the error could have come from that. So you did well to ask.
I hope you find the issue, your project looks interesting.

**Comment by AliYoussef97** [comment]
@gleize Thank you so much, I got it now!

Really sorry I had one last question, I understand from previous issues and the FAQ, that the scaling factor of 1.41 ($\sqrt{2}$) is for legacy reasons and only affects double-softmax matching, but is it possible to understand the reason behind bringing the descriptors to a hypersphere of radius 2? I have noticed that in the paper the temperature was $20^{-1}$ (0.05), but the default training configuration in the repo is $10^{-1}$ (0.1). I've read on of your previous [comments](https://github.com/facebookresearch/silk/issues/27#issuecomment-1637215158), that choosing a scale factor of 1 and a temperature of 0.1, would be equivalent to 1.41 and 0.05, but I fail to understand the significance of scaling the descriptors by ($\sqrt{2}$).

Sorry for asking to many quesitons.

**Comment by gleize** [comment]
Hi @AliYoussef97,

> [...] is it possible to understand the reason behind bringing the descriptors to a hypersphere of radius 2?

There are no reason other than it was originally used as a way to parametrize the temperature. Then later, I added the temperature parameter to the softmax formulation, which ended up with this unnecessary double parametrization (and I forgot to reset the initial $\sqrt{2}$ to 1). I left it because I didn't want to potentially break something up right before the submission.
